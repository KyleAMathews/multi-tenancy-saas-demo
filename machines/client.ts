import { v4 as uuidv4 } from "uuid"

export async function createRequest({ doc, mutator, request }) {
  const requests = doc.getArray(`requests`)
  const requestId = uuidv4()
  let resolveFunc, rejectFunc

  const promise = new Promise((resolve, reject) => {
    // Storing the resolve and reject functions for later use.
    resolveFunc = resolve
    rejectFunc = reject
  })
  function observe(event) {
    const inserted = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `insert`
    )
    const state = inserted.insert[0]
    // TODO only do this in case of user-directed timeout or network
    // disconnect errors e.g. normally we're fine just waiting to go online
    // but the app might want to error immediately if we disconnect or are offline.
    // if (state.error) {
    // requests.unobserve(observe)
    // return rejectFunc(state)
    // }
    if (state.done) {
      requests.unobserve(observe)
      return resolveFunc(state)
    }
  }

  requests.observe(observe)
  requests.push([
    {
      mutator,
      value: `requested`,
      id: requestId,
      done: false,
      clientCreate: new Date().toJSON(),
      request,
      response: {},
    },
  ])

  return promise
}
