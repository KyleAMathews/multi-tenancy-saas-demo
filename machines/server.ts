import * as Y from "yjs"

export function listen({ doc, serverConfig }) {
  const requests = doc.getArray(`requests`)
  requests.observe(async (event) => {
    const inserted = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `insert`
    )
    const retain = event.changes.delta.find(
      (item) => Object.keys(item)[0] === `retain`
    )
    const itemArray = retain?.retain || 0
    const state = event.target.get(itemArray)
    if (state.done !== true) {
      const mutatorFunc = await serverConfig.mutators[state.mutator]({
        state,
        doc,
      })
      doc.transact(() => {
        state.response = mutatorFunc()

        if (typeof state.response?.error !== `undefined`) {
          state.error = true
        } else {
          state.error = false
        }

        state.value = `responded`
        state.done = true
        state.serverResponded = new Date().toJSON()
        requests.delete(itemArray, 1)
        requests.insert(itemArray, [state])
      })
    }
  })
}
