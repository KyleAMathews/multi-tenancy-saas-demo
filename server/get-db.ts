import path from "path"
import Database from "libsql"

export function getDb(baseDir, dbName) {
  const dbPath = path.join(baseDir, `${dbName}.db`)
  const db = new Database(dbPath)

  return db
}
