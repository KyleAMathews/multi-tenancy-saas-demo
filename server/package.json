{
  "name": "server",
  "module": "index.ts",
  "type": "module",
  "devDependencies": {
    "bun-types": "latest",
    "tsx": "^3.13.0",
    "vitest": "^0.34.5"
  },
  "scripts": {
    "dev": "tsx index.ts",
    "build": "cd ../app-admin && npm run build && cp -r dist ../server/",
    "start": "NODE_ENV=production tsx index.ts",
    "test": "vitest"
  },
  "peerDependencies": {
    "typescript": "^5.0.0"
  },
  "dependencies": {
    "@coffeeandfun/google-profanity-words": "^2.1.0",
    "@libsql/client": "^0.3.5",
    "cors": "^2.8.5",
    "express": "^4.18.2",
    "fs-extra": "^11.1.1",
    "libsql": "^0.1.23",
    "node-sql-parser": "^4.11.0",
    "situated": "^0.0.1",
    "ws": "^8.14.2",
    "yjs": "^13.6.8"
  }
}
