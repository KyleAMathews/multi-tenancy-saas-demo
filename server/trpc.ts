import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"
import Parser from "node-sql-parser"
import { createClient } from "@libsql/client"
import * as util from "node:util"
import * as child_process from "node:child_process"
import { mapResultSet } from "./map-sqlite-resultset"
import { ProfanityEngine } from "@coffeeandfun/google-profanity-words"
const profanity = new ProfanityEngine()
console.log(Parser)
const parser = new Parser.Parser()

const execAsync = util.promisify(child_process.exec)

/**
 * Initialization of tRPC backend
 * Should be done only once per backend!
 */
const t = initTRPC.create()
/**
 * Export reusable router and procedure helpers
 * that can be used throughout the router
 */
const router = t.router
const publicProcedure = t.procedure

async function setupDb(db) {
  await db.execute(`CREATE TABLE Todo (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  title TEXT NOT NULL,
  completed BOOLEAN NOT NULL CHECK (completed IN (0, 1))
);`)
  await db.execute(
    `INSERT INTO Todo (title, completed) VALUES ('Go to Grocery Store', 0)`
  )
}

export const appRouter = router({
  ping: publicProcedure.mutation(async () => {
    return {
      transact: () => {
        true
      },
    }
  }),
  createDb: publicProcedure
    .input(
      z.object({
        name: z
          .string()
          .regex(/^[a-zA-Z0-9-_]{1,27}$/, {
            message: `App name must not contain spaces or special characters`,
          })
          .max(28, { message: `App name must be 28 characters or less` })
          .min(3),
        fromDb: z.string().min(3).optional(),
      })
    )
    .mutation(async function ({ input, ctx: { doc, adminDb, transact } }) {
      const dbs = doc.getMap(`dbs`)
      if (dbs.has(input.name)) {
        throw new TRPCError({
          code: `CONFLICT`,
          message: `A db by this name already exists`,
        })
      }

      const isProfane = await profanity.hasCurseWords(
        input.name.split(`-`).join(` `).split(`_`).join(` `)
      )
      if (isProfane) {
        throw new TRPCError({
          code: `BAD_REQUEST`,
          message: `Profane db names are now allowed.`,
        })
      }

      let createOutput
      let urlOutput
      let tokenOutput
      dbs.set(input.name, {
        name: input.name,
        state: `INITIALIZING`,
        updatedAt: new Date().toJSON(),
      })
      let command = `turso db create --group demo-multi-tenant-saas ${input.name}`
      if (input.fromDb) {
        command += ` --from-db=${input.fromDb}`
      }
      try {
        createOutput = await execAsync(command)
        dbs.set(input.name, {
          name: input.name,
          state: `CREATED`,
          updatedAt: new Date().toJSON(),
        })
        urlOutput = await execAsync(`turso db show ${input.name} --url`)
        tokenOutput = await execAsync(`turso db tokens create ${input.name}`)
        dbs.set(input.name, {
          name: input.name,
          state: `CREATING TABLES`,
          updatedAt: new Date().toJSON(),
        })
      } catch (e) {
        console.log(e)
        dbs.delete(input.name)
        throw new TRPCError({
          code: `INTERNAL_SERVER_ERROR`,
          message: `Error creating database or tables`,
          cause: e,
        })
      }
      console.log({ createOutput, urlOutput, tokenOutput })

      const url = urlOutput.stdout.trim()
      const authToken = tokenOutput.stdout.trim()
      console.log({ url, authToken })

      const db = createClient({ url, authToken })

      // Don't need to do setup for cloned dbs.
      if (!input.fromDb) {
        await setupDb(db)
      }
      // Async work first and then return func w/ any sync changes.
      // Validate db doesn't exist in both yjs and on disk.
      // Then create db and create table.
      //
      // TODO also a test to run queries.
      const updatedAt = new Date().toJSON()
      await adminDb.execute({
        sql: `INSERT INTO dbs values (:url, :authToken, :state, :name, :updatedAt)`,
        args: {
          url,
          authToken,
          state: `READY`,
          name: input.name,
          updatedAt,
        },
      })

      const totals = mapResultSet(
        await db.execute(
          `select completed, count(*) as count from Todo group by completed`
        )
      )
      console.log({ totals })

      transact(() => {
        dbs.set(input.name, {
          url,
          authToken,
          state: `READY`,
          name: input.name,
          total: totals.map((row) => row.count).reduce((a, b) => a + b, 0),
          completed: totals.find((row) => row.completed === 1)?.count || 0,
          updatedAt,
        })
      })

      return { url, name: input.name }
      // return function () {
      // dbs.set(input.name, {
      // url,
      // authToken,
      // state: `READY`,
      // name: input.name,
      // total: totals.map((row) => row.count).reduce((a, b) => a + b, 0),
      // completed: totals.find((row) => row.completed === 1)?.count || 0,
      // updatedAt,
      // })
      // return { url, name: input.name }
      // }
    }),
  deleteDb: publicProcedure
    .input(z.object({ name: z.string() }))
    .mutation(async function ({ input, ctx: { doc, adminDb, transact } }) {
      const dbs = doc.getMap(`dbs`)
      if (!dbs.has(input.name)) {
        return function () {
          return {
            error: `DB with the name ${input.name} doesn't exists`,
          }
        }
      }

      let destroyOutput
      try {
        destroyOutput = await execAsync(`turso db destroy ${input.name} --yes`)
      } catch (e) {
        console.log(e)
        return function () {
          return { error: e }
        }
      }

      await adminDb.execute({
        sql: `DELETE FROM dbs WHERE name=:name`,
        args: {
          name: input.name,
        },
      })

      transact(() => {
        dbs.delete(input.name)
      })

      return input.name
    }),
  selectDb: publicProcedure
    .input(z.object({ name: z.string(), sql: z.string() }))
    .query(async function ({ input, ctx: { doc } }) {
      try {
        const dbInfo = doc.getMap(`dbs`).get(input.name)
        const db = createClient({
          url: dbInfo.url,
          authToken: dbInfo.authToken,
        })
        const ast = parser.astify(input.sql)
        if (ast.type === `select`) {
          const results = mapResultSet(await db.execute(input.sql))
          // Async work first and then return func w/ any sync changes.
          return {
            ok: true,
            results,
          }
        } else {
          throw new TRPCError({
            code: `BAD_REQUEST`,
            message: `Only select operations are allowed`,
          })
        }
      } catch (e) {
        console.log(e)
        throw new TRPCError({
          code: `INTERNAL_SERVER_ERROR`,
          cause: e,
        })
      }
    }),
})
// userUpdateName: publicProcedure
// .input(z.object({ id: z.string(), name: z.string() }))
// .mutation(async (opts) => {
// const {
// input,
// ctx: { users },
// } = opts
// let user
// let id
// users.forEach((u, i) => {
// if (u.id === input.id) {
// user = u
// id = i
// }
// })
// const newUser = { ...user, name: input.name }
// return {
// mutations: () => {
// users.delete(id, 1)
// users.insert(id, [newUser])
// },
// response: newUser,
// }
// }),
// })

export type AppRouter = typeof appRouter
