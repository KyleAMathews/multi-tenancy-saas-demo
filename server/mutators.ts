import path from "path"
import Parser from "node-sql-parser"
import fs from "fs-extra"
import { getDb } from "./get-db"
const parser = new Parser.Parser()

async function setupDb(db) {
  db.exec(`CREATE TABLE Todo (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  title TEXT NOT NULL,
  completed BOOLEAN NOT NULL CHECK (completed IN (0, 1))
);`)
  db.exec(
    `INSERT INTO Todo (title, completed) VALUES ('Go to Grocery Store', 0)`
  )
}
function deleteDb(dbPath) {
  return fs.unlinkSync(dbPath)
}

export const serverConfig = ({ dbsDir }) => {
  return {
    mutators: {
      ping: async function ({ state, doc }) {
        return function () {
          return { ok: true, request: state.request }
        }
      },
      createDb: async function ({ state, doc }) {
        // Validate app/db name
        const regex = /^[a-zA-Z0-9-_]{1,27}$/
        if (!regex.test(state.request.name)) {
          console.log({
            name: state.request.name,
            test: regex.test(state.request.name),
          })
          return function () {
            return {
              error: `App names must not contain spaces or special characters and be less than 28 characters`,
            }
          }
        }

        const dbs = doc.getMap(`dbs`)
        if (dbs.has(state.request.name)) {
          return function () {
            return { error: `DB Already exists` }
          }
        }
        const dbPath = path.join(dbsDir, `${state.request.name}.db`)
        if (fs.existsSync(dbPath)) {
          return function () {
            return {
              error: `DB Already exists on disk (though oddly not in the map)`,
            }
          }
        }

        const db = getDb(dbsDir, state.request.name)
        await setupDb(db)
        // Async work first and then return func w/ any sync changes.
        // Validate db doesn't exist in both yjs and on disk.
        // Then create db and create table.
        //
        // TODO also a test to run queries.
        return function () {
          dbs.set(state.request.name, {
            dbPath,
            name: state.request.name,
            total: 1,
            completed: 0,
          })
          return { dbPath, name: state.request.name }
        }
      },
      deleteDb: async function ({ state, doc }) {
        const dbs = doc.getMap(`dbs`)
        if (!dbs.has(state.request.name)) {
          return function () {
            return {
              error: `DB with the name ${state.request.name} doesn't exists`,
            }
          }
        }
        const dbPath = path.join(dbsDir, `${state.request.name}.db`)
        if (!fs.existsSync(dbPath)) {
          return function () {
            return {
              error: `DB doesn't exist on disk`,
            }
          }
        }

        deleteDb(dbPath)
        return function () {
          dbs.delete(state.request.name)
          return { dbPath, name: state.request.name }
        }
      },
      selectDb: async function ({ state }) {
        try {
          const db = getDb(dbsDir, state.request.name)
          const ast = parser.astify(state.request.sql)
          if (ast.type === `select`) {
            console.log(`query`, state.request.sql)
            let results
            results = db.prepare(state.request.sql).all()
            // Async work first and then return func w/ any sync changes.
            return function () {
              return { ok: true, results }
            }
          } else {
            return function () {
              return { error: `Only select operations are allowed` }
            }
          }
        } catch (e) {
          console.log(e)
          return function () {
            return { error: e }
          }
        }
      },
    },
  }
}
