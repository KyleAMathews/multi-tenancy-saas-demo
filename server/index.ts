import cors from "cors"
import path from "path"
import express from "express"
import { WebSocketServer } from "ws"
import { fileURLToPath } from "url"
import fs from "fs-extra"
import { setupWSConnection, getYDoc } from "situated"
import listen from "../machines/server"
import mapResultSet from "../map-sqlite-resultset"
console.log({ mapResultSet })
import Parser from "node-sql-parser"
import { serverConfig } from "./mutators"
import { createClient } from "@libsql/client"

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const baseDir =
  process.env.BASE_DATA_DIR || path.resolve(process.cwd(), `.cache`)
const file = path.join(baseDir, `db.json`)
const dbsDir = path.join(baseDir, `dbs`)
console.log({ dbsDir, dir: path.dirname(dbsDir) })
fs.ensureDirSync(path.dirname(file))
fs.ensureDirSync(dbsDir)

const app = express()
app.use(express.json())
app.use(
  cors({
    origin: `http://localhost:5174`,
    credentials: true,
  })
)

// Serve static assets.
app.use(`/`, express.static(path.join(__dirname, `../dist`)))

// handle every other route with index.html, which will contain
// a script tag to your application's JavaScript file(s).
app.get(`*`, function (request, response) {
  response.sendFile(path.resolve(__dirname, `../dist/index.html`))
})

app.post(`/invalidate/:dbName`, async (req, res) => {
  const dbName = req.params.dbName

  const doc = getYDoc(`app-doc`)
  const dbs = doc.getMap(`dbs`)

  if (dbs.has(dbName)) {
    const ydocDb = dbs.get(dbName)
    // Query for total + completed and set.
    const db = createClient({ url: ydocDb.url, authToken: ydocDb.authToken })
    const totals = mapResultSet.mapResultSet(
      await db.execute(
        `select completed, count(*) as count from todo group by completed`
      )
    )
    console.log({ ydocDb, totals })
    doc.transact(() => {
      ydocDb.total = totals.map((row) => row.count).reduce((a, b) => a + b, 0)
      ydocDb.completed = totals.find((row) => row.completed === 1)?.count || 0
      ydocDb.updatedAt = new Date().toJSON()
      dbs.set(dbName, ydocDb)
    })
    console.log({ totals })
    res.send(`ok`)
  }
})

const wsServer = new WebSocketServer({ noServer: true })
wsServer.on(`connection`, setupWSConnection)

let port = 3000
if (process.env.NODE_ENV === `production`) {
  port = 4000
}

const server = app.listen(port, () => {
  console.log(`API listening on port ${port}`)
  const doc = getYDoc(`app-doc`)
  console.log(`got doc`)
  listen.listen({ doc, serverConfig: serverConfig({ dbsDir }) })
})

server.on(`upgrade`, (request, socket, head) => {
  wsServer.handleUpgrade(request, socket, head, (socket) => {
    wsServer.emit(`connection`, socket, request)
  })
})
