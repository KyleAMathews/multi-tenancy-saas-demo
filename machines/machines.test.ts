import { expect, test } from "bun:test"
import * as Y from "yjs"
import fs from "fs"
import path from "path"
import Database from "libsql"
const { Parser } = require(`node-sql-parser`)
const parser = new Parser()

import { listen } from "./server"
import { createRequest } from "./client"

test(`update a robot name`, async () => {
  // const { machine } = require(`./ping-pong`)

  const doc = new Y.Doc()

  const robots = doc.getMap(`robots`)
  const robotId = `123`
  robots.set(robotId, { id: robotId, name: `boop` })

  // Config that gets new requests & calls right function.
  const serverConfig = {
    mutators: {
      updateRobotName: async function ({ state, doc }) {
        await new Promise((resolve) => setTimeout(resolve, 3))
        // Async work first and then return func w/ any sync changes.
        return function () {
          const robots = doc.getMap(`robots`)
          const robot = robots.get(state.request.id)
          robot.name = state.request.name
          robots.set(state.request.id, robot)
          return { ok: true }
        }
      },
    },
  }

  listen({ doc, serverConfig })

  // Server instantiation
  // listen(doc, serverConfig)
  //
  // client instantiation
  // hmm just call createRequest({doc, mutator, request})

  // Client code
  // TODO
  // - create an id so can await it coming back — haha there's a generic
  // state machine for a client to run a client/server state machine. You pass in
  // machine and the initial state and then off it goes and you just await
  // it finishing. That handles creating a uuid and awaiting it to come back
  // updated.
  //
  // Or I guess it's a sort of abstraction for "I advanced this state machine as far as I can
  // and now I'm tossing it to the network for the next node to change". So await a change to the machine.
  //
  // Run a state machine until you hit a point that you want to serialize and it for
  // running on the server and then pick it back up again when it comes back.
  //
  // Basically a job though I guess mutation is more generic.
  //
  // I'm not sure xstate is needed here — just do what I did in the post really...
  //
  // xstate would be great for more complex stuff but probably just an object w/ some
  // validation is enough.
  //
  //
  // TODO add helper function to generate the request & await the response.
  // Server function which does the mutation on the doc.
  // switch to updateName
  let newName = `beep`
  const requestObject = await createRequest({
    doc,
    mutator: `updateRobotName`,
    request: { id: robotId, name: newName },
  })

  const { id, clientCreate, serverResponded, ...toSnapshot } = requestObject
  expect(toSnapshot).toMatchSnapshot()
  expect(requestObject.done).toBeTrue()
  expect(robots.get(requestObject.request.id).name).toEqual(newName)

  newName = `boop`
  const requestObject2 = await createRequest({
    doc,
    mutator: `updateRobotName`,
    request: { id: robotId, name: newName },
  })

  expect(robots.get(requestObject2.request.id).name).toEqual(newName)
})

test(`create/delete dbs`, async () => {
  const doc = new Y.Doc()

  const dbs = doc.getMap(`dbs`)

  // Generate a random directory name
  const dirName =
    `test-` + Date.now() + `-` + Math.random().toString(36).substring(2, 7)
  const dirPath = path.join(`/tmp`, dirName)

  try {
    // Create the directory synchronously
    fs.mkdirSync(dirPath)
  } catch (err) {
    console.error(`Failed to create directory:`, err)
  }

  async function setupDb(dbPath) {
    const db = new Database(dbPath)
    db.exec(
      `CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, email TEXT)`
    )
    db.exec(
      `INSERT INTO users (id, name, email) VALUES (1, 'Alice', 'alice@example.org')`
    )
  }

  // Config that gets new requests & calls right function.
  const serverConfig = {
    mutators: {
      createDb: async function ({ state, doc }) {
        const dbs = doc.getMap(`dbs`)
        if (dbs.has(state.request.name)) {
          return function () {
            return { error: `DB Already exists` }
          }
        }
        const dbPath = path.join(dirPath, `${state.request.name}.db`)
        if (fs.existsSync(dbPath)) {
          return function () {
            return {
              error: `DB Already exists on disk (though oddly not in the map)`,
            }
          }
        }

        await setupDb(dbPath)
        // Async work first and then return func w/ any sync changes.
        // Validate db doesn't exist in both yjs and on disk.
        // Then create db and create table.
        //
        // TODO also a test to run queries.
        return function () {
          dbs.set(state.request.name, { foo: true })
          return { dbPath }
        }
      },
      deleteDb: async function ({ state, doc }) {
        const dbPath = path.join(dirPath, `${state.request.name}.db`)
        fs.unlinkSync(dbPath)
        // Async work first and then return func w/ any sync changes.
        return function () {
          const dbs = doc.getMap(`dbs`)
          dbs.delete(state.request.name)
          return { ok: true }
        }
      },
      selectDb: async function ({ state, doc }) {
        const dbPath = path.join(dirPath, `${state.request.name}.db`)
        const db = new Database(dbPath)
        const ast = parser.astify(state.request.sql)
        if (ast.type === `select`) {
          const results = db.prepare(state.request.sql).all()
          // Async work first and then return func w/ any sync changes.
          return function () {
            return { ok: true, results }
          }
        } else {
          return function () {
            return { error: `Only select operations are allowed` }
          }
        }
      },
    },
  }

  listen({ doc, serverConfig })

  // Test creating a db.
  const requestObject = await createRequest({
    doc,
    mutator: `createDb`,
    request: { name: `foo` },
  })

  expect(dbs.get(`foo`)).toMatchSnapshot()

  // Test duplicate returns an error.
  const dupRequestObject = await createRequest({
    doc,
    mutator: `createDb`,
    request: { name: `foo` },
  })

  expect(dupRequestObject.error).toBeTrue()

  // Test running a valid and then invalid query.
  const selectReq = await createRequest({
    doc,
    mutator: `selectDb`,
    request: { name: `foo`, sql: `select * from users` },
  })

  expect(selectReq.response.results).toMatchSnapshot()

  const insertReq = await createRequest({
    doc,
    mutator: `selectDb`,
    request: {
      name: `foo`,
      sql: `INSERT INTO users (id, name, email) VALUES (2, 'Alice', 'alice@example.org')`,
    },
  })

  expect(insertReq.error).toBeTrue()

  // Test deleting a db.
  await createRequest({
    doc,
    mutator: `deleteDb`,
    request: { name: `foo` },
  })

  expect(dbs.has(`foo`)).toBeFalse()
  expect(fs.existsSync(requestObject.response.dbPath)).toBeFalse()
})
