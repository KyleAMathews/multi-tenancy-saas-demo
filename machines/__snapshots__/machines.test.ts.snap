// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`update a robot name 1`] = `
{
  "done": true,
  "error": false,
  "mutator": "updateRobotName",
  "request": {
    "id": "123",
    "name": "beep",
  },
  "response": {
    "ok": true,
  },
  "value": "responded",
}
`;

exports[`create/delete dbs 1`] = `
{
  "foo": true,
}
`;

exports[`create/delete dbs 2`] = `
[
  {
    "email": "alice@example.org",
    "id": 1,
    "name": "Alice",
  },
]
`;
